DataStructures/AATree/AATree.cs

```diff
@@ -0,0 +1,587 @@
+            }
+        }
+
+        /// <summary>
+        /// Recursive to find the maximum/right-most element.
+        /// </summary>
+        /// <param name="node">The node to traverse from.</param>
+        /// <returns>The node with the maximum/right-most element.</returns>
+        private AATreeNode<TKey> GetMax(AATreeNode<TKey> node)
+        {
+            if (node.Right == null)
+            {
+                return node;
+            }
+            else
+            {
+                return GetMax(node.Right);
+            }
+        }
+
+        /// <summary>
+        /// Recursive to find the minimum/left-most element.
+        /// </summary>
+        /// <param name="node">The node to traverse from.</param>
+        /// <returns>The node with the minimum/left-most element.</returns>
+        private AATreeNode<TKey> GetMin(AATreeNode<TKey> node)
+        {
+            if (node.Left == null)
+            {
+                return node;
+            }
+            else
+            {
+                return GetMin(node.Left);
+            }
+        }
+
+        /// <summary>
+        /// Checks various properties to determine if the tree is a valid AA Tree.
+        /// Throws exceptions if properties are violated.
+        /// Useful for debugging.
+        /// </summary>
+        /// <remarks>
+        /// The properties that are checked are:
+        /// <list type="number">
+        /// <item>The level of every leaf node is one.</item>
+        /// <item>The level of every left child is exactly one less than that of its parent.</item>
+        /// <item>The level of every right child is equal to or one less than that of its parent.</item>
+        /// <item>The level of every right grandchild is strictly less than that of its grandparent.</item>
+        /// <item>Every node of level greater than one has two children.</item>
+        /// </list>
+        /// More information: https://en.wikipedia.org/wiki/AA_tree .
+        /// </remarks>
+        /// <param name="node">The node to check from.</param>
+        /// <returns>true if node passes all checks, false otherwise.</returns>
+        private bool ValidateTree(AATreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return true;
+            }
+
+            // Check level == 1 if node if a leaf node.
+            bool leafNodeCheck = CheckLeafNode(node);
+
+            // Check level of left child is exactly one less than parent.
+            bool leftCheck = CheckLeftSubtree(node);
+
+            // Check level of right child is equal or one less than parent.
+            bool rightCheck = CheckRightSubtree(node);
+
+            // Check right grandchild level is less than node.
+            bool grandchildCheck = CheckRightGrandChild(node);
+
+            // Check if node has two children if not leaf.
+            bool nonLeafChildrenCheck = CheckNonLeafChildren(node);
+
+            bool thisNodeResult = leafNodeCheck &&
```

Comment:
Codacy found an issue: [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6367315)